### PR TITLE
feat(pie-notification): DSW-1351 Add user actions

### PR DESCRIPTION
**Pull request by thejfreitas**
## Describe your changes (can list changeset entries if preferable)

In this PR I am adding the user actions for `pie-notification` such as:

```
isDismissible
leadingAction
supportingAction
hasStackedActions
```
As well support for custom events.

### Screenshots

Screenshot 2024-02-20 at 4 45 25 PM

Screenshot 2024-02-23 at 4 52 57 PM

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests
- [x] If it is a `PIE Docs` change, I have reviewed the Docs site preview
- [x] If it is a component change, I have reviewed the Storybook preview
- [x] If there are visual test updates, I have reviewed them properly before approving

### CODE REVIEW

File: .changeset/flat-adults-trade.md
Comment (raoufswe) [inline .changeset/flat-adults-trade.md:7]: i think the changeset might be an old one 👀

Comment (thejfreitas) [reply .changeset/flat-adults-trade.md:7]: it ts from the latest PR, and I just added a new one containing the changes of this PR

File: packages/components/pie-notification/src/index.ts
Comment (raoufswe) [inline packages/components/pie-notification/src/index.ts:136]: we use camelCasing for custom props

```suggestion
            <footer class="${componentClass}-footer" data-test-id="${componentSelector}-footer" isCompact="${this.isCompact}" isStacked="${this.hasStackedActions && !this.isCompact}">
```

File: packages/components/pie-notification/test/component/pie-notification.spec.ts
Comment (raoufswe) [inline packages/components/pie-notification/test/component/pie-notification.spec.ts:48]: this looks redundant since we have it already on line 47

File: .changeset/flat-adults-trade.md
Comment (raoufswe) [reply .changeset/flat-adults-trade.md:7]: let's delete it then if it was part of the previous PR

File: packages/components/pie-notification/src/index.ts
Comment (raoufswe) [inline packages/components/pie-notification/src/index.ts:36]: do you think we can move it to "./defs"?

File: packages/components/pie-notification/src/index.ts
Comment (raoufswe) [reply packages/components/pie-notification/src/index.ts:136]: also i think we should use [nothing](https://lit.dev/docs/api/templates/#nothing) is isStacked as it provides a consistent behavior between various expression binding contexts.

```suggestion
            <footer class="${componentClass}-footer" data-test-id="${componentSelector}-footer" is-compact="${this.isCompact}" is-stacked="${this.hasStackedActions && !this.isCompact ? true : nothing}">
```

File: packages/components/pie-notification/src/index.ts
Comment (raoufswe) [inline packages/components/pie-notification/src/index.ts:312]: i think we should use type="submit" here since the parent element isn't a form. 
do you think we can use type="button" instead?

File: packages/components/pie-notification/test/component/pie-notification.spec.ts
Comment (thejfreitas) [reply packages/components/pie-notification/test/component/pie-notification.spec.ts:48]: 🤔 I believe this was git merge getting confused. I will remove it.

**Comment by raoufswe** [comment]
a side note, seems like the icons don't get updated when you change the variant and they default to the default info icon.

File: packages/components/pie-notification/src/index.ts
Comment (raoufswe) [reply packages/components/pie-notification/src/index.ts:136]: nit: a small suggestion, we don't need to pass the `leadingAction` and supportingAction as args to `renderFooter` since they're still part of the same class and can be accessed directly using the `this` keyword. 

the same thing applies for `renderNotificationHeading` and other a few places where we pass args around. happy to address that in a follow up PR if you find it a bit clearer.

**Comment by thejfreitas** [comment]
> a side note, seems like the icons don't get updated when you change the variant and they default to the default info icon.

@raoufswe There's an icon being passed as named slot on `pie-notification` storybook, if you comment that icon slot you will be able to see the default icons for the designated variant.

File: packages/components/pie-notification/src/index.ts
Comment (thejfreitas) [reply packages/components/pie-notification/src/index.ts:36]: Actually this should not exists anymore.

File: packages/components/pie-notification/src/index.ts
Comment (thejfreitas) [reply packages/components/pie-notification/src/index.ts:312]: You're totally right!

File: packages/components/pie-notification/src/index.ts
Comment (thejfreitas) [reply packages/components/pie-notification/src/index.ts:132]: I found a better way to use camelCasing and avoiding ternaries by writing the custom props this way:
```
<div ?isCompact="${this.isCompact}" ?isStacked="${this.hasStackedActions && !this.isCompact}">
```
About your suggestion of removing args, I would like to bring this topic to the rest of the team.

File: apps/pie-storybook/stories/pie-notification.stories.ts
Comment (jamieomaguire) [inline apps/pie-storybook/stories/pie-notification.stories.ts:27]: We could leverage storybook actions here instead of console logs - entirely up to you but there should be examples in the `pie-input` stories if you're interested

File: packages/components/pie-notification/test/component/pie-notification.spec.ts
Comment (siggerzz) [inline packages/components/pie-notification/test/component/pie-notification.spec.ts:252]: These tests should probably be visual tests rather than component tests. If you need a hand setting these up, please reach out!

Comment (thejfreitas) [reply packages/components/pie-notification/test/component/pie-notification.spec.ts:252]: @siggerzz I just created visual tests adding based on the unit tests. I think maintaining both visual and unit tests enriches the test coverage.

File: packages/components/pie-notification/test/visual/pie-notification.spec.ts
Comment (raoufswe) [inline packages/components/pie-notification/test/visual/pie-notification.spec.ts:117]: i think this test can be combined with the main first test.
do you think we can pass headingLevel: headingLevels to line 28?

File: apps/pie-storybook/stories/pie-notification.stories.ts
Comment (raoufswe) [inline apps/pie-storybook/stories/pie-notification.stories.ts:146]: do we need to umcomment this?

File: packages/components/pie-notification/test/visual/pie-notification.spec.ts
Comment (thejfreitas) [reply packages/components/pie-notification/test/visual/pie-notification.spec.ts:117]: @raoufswe I am not sure. Percy cannot handle several screenshots at once, by adding this prop we'll generate a lot of screenshots that looks the same due the combination of other props together.

File: apps/pie-storybook/stories/pie-notification.stories.ts
Comment (thejfreitas) [reply apps/pie-storybook/stories/pie-notification.stories.ts]: Sure thing

File: packages/components/pie-notification/test/visual/pie-notification.spec.ts
Comment (raoufswe) [reply packages/components/pie-notification/test/visual/pie-notification.spec.ts:117]: aha makes sense

**Review by raoufswe — DISMISSED** [review]

File: packages/components/pie-notification/src/defs.ts
Comment (jamieomaguire) [inline packages/components/pie-notification/src/defs.ts:20]: I am a bit unsure about this. Until now we have not accepted callback functions in our props because we want to move away from React patterns and stick to event based solutions. I.e. a component emits an event, the consumer chooses how to respond to it in their own code.

Can I ask why we have chosen to go with the callback approach here?

File: .changeset/flat-adults-trade.md
Comment (xander-marjoram) [reply .changeset/flat-adults-trade.md:7]: Looks like this file still needs to be deleted as it was included in the previous PR: https://github.com/justeattakeaway/pie/pull/1254/files#diff-dd3a122036d1313becc5980f175cc80d41b11f46c142fb22a3997be057a67455

File: packages/components/pie-notification/src/defs.ts
Comment (xander-marjoram) [reply packages/components/pie-notification/src/defs.ts:20]: I agree with @jamieomaguire here. We have a good example in the modal component of dispatching custom events which consumers can then listen for:

https://github.com/justeattakeaway/pie/blob/main/packages/components/pie-modal/src/index.ts#L225

Comment (thejfreitas) [reply packages/components/pie-notification/src/defs.ts]: @xander-marjoram @jamieomaguire After today's call I finally understand how this event based approach works and I did some refactor in the component and now everything is working properly as well onClick has been removed out of this implementation. Thanks Jamie to take time to explain to me .

File: apps/pie-storybook/stories/pie-notification.stories.ts
Comment (jamieomaguire) [inline apps/pie-storybook/stories/pie-notification.stories.ts:148]: 🔥

File: packages/components/pie-notification/src/index.ts
Comment (jamieomaguire) [inline packages/components/pie-notification/src/index.ts:41]: 🔥

File: packages/components/pie-notification/test/visual/pie-notification.spec.ts
Comment (jamieomaguire) [inline packages/components/pie-notification/test/visual/pie-notification.spec.ts:97]: I think this can be removed now

File: .changeset/flat-adults-trade.md
Comment (thejfreitas) [reply .changeset/flat-adults-trade.md:7]: This file has been deleted.

File: packages/components/pie-notification/test/visual/pie-notification.spec.ts
Comment (thejfreitas) [reply packages/components/pie-notification/test/visual/pie-notification.spec.ts]: Removed!

**Review by jamieomaguire — DISMISSED** [review]

**Review by raoufswe — DISMISSED** [review]

**Review by siggerzz — APPROVED** [review]

**Comment by siggerzz** [comment]
Hey Jose!

Looking good, however I've noticed a bit of overlap on Percy in Firefox, would it be possible to get this addressed before we merge please?

Screenshot 2024-03-22 at 11 02 59

Screenshot 2024-03-22 at 11 06 39

**Review by siggerzz — CHANGES_REQUESTED** [review]
Just a few comments regarding the overlapping issues on Firefox, but happy to approve once this is resolved!

**Review by siggerzz — COMMENTED** [review]
Hey Jose!

Looking good, however I've noticed a bit of overlap on Percy in Firefox.

If it's not trivial to fix as part of this PR, a follow-up PR after this once has been merged would be much appreciated!

Screenshot 2024-03-22 at 11 02 59

Screenshot 2024-03-22 at 11 06 39

**Review by siggerzz — APPROVED** [review]
Thanks for addressing the Firefox issues with the close icon @thejfreitas!

Looks like there's still a bit of overlap for the icon in Firefox, but I'm happy for this to be addressed in a future PR 👍🏼 
Screenshot 2024-03-26 at 10 11 38

**Review by jamieomaguire — APPROVED** [review]

**Pull request merged**
